File: CS/mailClient.cs
﻿using System;
using System.IO;
using System.Net.Sockets;
using System.Net.Security;
using System.Text;
using System.Windows.Forms;

class mailClient
{
    private string SMTP_SERVER;
    private int SMTP_PORT;

    public mailClient(string host) {
        SMTP_SERVER = host;
        SMTP_PORT = 465;
    }
    /*
    static void Main()
    {
        string smtpId = "smtpID"; // NAVER ID
        string password = "password"; // NAVER 비밀번호
        string senderEmail = "example@naver.com"; // 발신자 이메일
        string recipient = "example@gmail.com"; // 수신자 이메일
        string subject = "Test Email"; // 이메일 제목
        string body = "Hello, this is a test email from C# using Naver SMTP."; // 이메일 본문

        SendEmail(smtpId, password, senderEmail, recipient, subject, body);
    }
    */

    public void SendEmail(string senderEmail, string password, string recipient, string subject, string body)
    {
        try {
            using (TcpClient client = new TcpClient(SMTP_SERVER, SMTP_PORT))
            using (SslStream sslStream = new SslStream(client.GetStream(), false, new RemoteCertificateValidationCallback((sender, certificate, chain, sslPolicyErrors) => true))) {
                sslStream.AuthenticateAsClient(SMTP_SERVER);
                using (StreamReader reader = new StreamReader(sslStream))
                using (StreamWriter writer = new StreamWriter(sslStream) { AutoFlush = true }) {
                    // 서버 응답 읽기
                    ReadResponse(reader);

                    // EHLO
                    writer.WriteLine($"EHLO {SMTP_SERVER}");
                    ReadResponse(reader);

                    // AUTH PLAIN
                    string authString = Convert.ToBase64String(Encoding.UTF8.GetBytes($"\0{senderEmail}\0{password}"));
                    writer.WriteLine($"AUTH PLAIN {authString}");
                    ReadResponse(reader);

                    // MAIL FROM
                    writer.WriteLine($"MAIL FROM:<{senderEmail}>");
                    ReadResponse(reader);

                    // RCPT TO
                    writer.WriteLine($"RCPT TO:<{recipient}>");
                    ReadResponse(reader);

                    // DATA
                    writer.WriteLine("DATA");
                    ReadResponse(reader);

                    // 이메일 본문 전송
                    StringBuilder emailContent = new StringBuilder();
                    emailContent.AppendLine($"From: {senderEmail}");
                    emailContent.AppendLine($"To: {recipient}");
                    emailContent.AppendLine($"Subject: {subject}");
                    emailContent.AppendLine();
                    emailContent.AppendLine(body);
                    emailContent.AppendLine(".");
                    writer.WriteLine(emailContent.ToString());
                    ReadResponse(reader);

                    // QUIT
                    writer.WriteLine("QUIT");
                    ReadResponse(reader);
                }
            }
        }
        catch (Exception ex) {
            MessageBox.Show("Error: " + ex.Message);
        }
    }

    static void ReadResponse(StreamReader reader)
    {
        string response = reader.ReadLine().Substring(0,3);
        if(response !="220" && response!="250" &&response!="354" &&response!="334" &&response!="235") {
            MessageBox.Show($"메일 발송 중 오류가 발생했습니다. {response}");
        }
    }
}

File: Java/src/MailClient.java


import javax.net.ssl.SSLSocket;
import javax.net.ssl.SSLSocketFactory;
import javax.swing.JOptionPane;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Base64;

public class MailClient {
    private String SMTP_SERVER;
    private int SMTP_PORT;

    public MailClient(String host) {
        this.SMTP_SERVER = host;
        this.SMTP_PORT = 465;
    }

    public void sendEmail(String senderEmail, String password, String recipient, String subject, String body) {
        SSLSocket socket = null;
        try {
            // SSL 소켓 생성 및 서버 연결
            SSLSocketFactory factory = (SSLSocketFactory) SSLSocketFactory.getDefault();
            socket = (SSLSocket) factory.createSocket(SMTP_SERVER, SMTP_PORT);
            socket.startHandshake();

            BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(socket.getOutputStream()));

            // 서버 응답 읽기
            readResponse(reader);

            // EHLO
            writer.write("EHLO " + SMTP_SERVER + "\r\n");
            writer.flush();
            readResponse(reader);

            // AUTH PLAIN
            String authString = "\0" + senderEmail + "\0" + password;
            String authBase64 = Base64.getEncoder().encodeToString(authString.getBytes("UTF-8"));
            writer.write("AUTH PLAIN " + authBase64 + "\r\n");
            writer.flush();
            readResponse(reader);

            // MAIL FROM
            writer.write("MAIL FROM:<" + senderEmail + ">\r\n");
            writer.flush();
            readResponse(reader);

            // RCPT TO
            writer.write("RCPT TO:<" + recipient + ">\r\n");
            writer.flush();
            readResponse(reader);

            // DATA
            writer.write("DATA\r\n");
            writer.flush();
            readResponse(reader);

            // 이메일 본문 전송
            StringBuilder emailContent = new StringBuilder();
            emailContent.append("From: ").append(senderEmail).append("\r\n");
            emailContent.append("To: ").append(recipient).append("\r\n");
            emailContent.append("Subject: ").append(subject).append("\r\n");
            emailContent.append("\r\n").append(body).append("\r\n.\r\n");
            writer.write(emailContent.toString());
            writer.flush();
            readResponse(reader);

            // QUIT
            writer.write("QUIT\r\n");
            writer.flush();
            readResponse(reader);

        } catch (Exception e) {
            e.printStackTrace();
            System.out.println("Error: " + e.getMessage());
            JOptionPane.showMessageDialog(null, e.getMessage(),"Warning",JOptionPane.WARNING_MESSAGE);
        } finally {
            // 소켓 닫기
            try {
                if (socket != null && !socket.isClosed()) {
                    socket.close();
                }
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        }
    }

    private void readResponse(BufferedReader reader) throws Exception {
        String response = reader.readLine().substring(0, 3);

        // 예상치 못한 응답 코드가 있을 경우 오류 메시지 표시
        if (!response.equals("220") && !response.equals("250") && !response.equals("354") &&
            !response.equals("334") && !response.equals("235") && !response.equals("221")) {
            throw new Exception("메일 발송 중 오류가 발생했습니다. 코드: " + response);
        }
    }
    /*
    public static void main(String[] args) {
    	/*
        String smtpId = "0603alswo"; // NAVER ID
        String password = "0603@alswo"; // NAVER 비밀번호
        String senderEmail = "0603alswo@naver.com"; // 발신자 이메일
        String recipient = "0603alswo@gmail.com"; // 수신자 이메일
        String subject = "Test Email"; // 이메일 제목
        String body = "Hello, this is a test email from Java using Naver SMTP."; // 이메일 본문

        MailClient mailClient = new MailClient("smtp.naver.com");
        mailClient.sendEmail(senderEmail, recipient, subject, body);
        
    }
*/
    
}
